### PR TITLE
[flang][cuda] Flag globals used in device function

### PR DESCRIPTION
**Pull request by clementval**
This patch adds a pass to flag globals used in device function with the device or constant attribute as they need to be available in the GPU module later on.

### CODE REVIEW

File: flang/lib/Optimizer/Transforms/CufImplicitDeviceGlobal.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+namespace fir {
+#define GEN_PASS_DEF_CUFIMPLICITDEVICEGLOBAL
+#include "flang/Optimizer/Transforms/Passes.h.inc"
+} // namespace fir
+
+namespace {
+
+static fir::GlobalOp getGlobalOpFromValue(mlir::Value v) {
+  if (auto addrOfOp{mlir::dyn_cast_or_null<fir::AddrOfOp>(v.getDefiningOp())}) {
+    auto sym{mlir::SymbolTable::lookupNearestSymbolFrom(
```
Comment (vzakhari) [inline flang/lib/Optimizer/Transforms/CufImplicitDeviceGlobal.cpp:28]: I would recommend building a `SymbolTable` out of the module and using it here. This should be faster, in general.

File: flang/lib/Optimizer/Transforms/CufImplicitDeviceGlobal.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  }
+}
+
+class CufImplicitDeviceGlobal
+    : public fir::impl::CufImplicitDeviceGlobalBase<CufImplicitDeviceGlobal> {
+public:
+  void runOnOperation() override {
+    auto *ctx = &getContext();
+    mlir::RewritePatternSet patterns(ctx);
+    mlir::ConversionTarget target(*ctx);
```
Comment (vzakhari) [inline flang/lib/Optimizer/Transforms/CufImplicitDeviceGlobal.cpp:60]:
```suggestion
```

File: flang/lib/Optimizer/Transforms/CufImplicitDeviceGlobal.cpp (hunk 1 above)
Comment (clementval) [reply flang/lib/Optimizer/Transforms/CufImplicitDeviceGlobal.cpp]: Great recommendation! Just updated it.

**Review by vzakhari — APPROVED** [review]
Thanks for the additional clean-up!

**Review by wangzpgi — APPROVED** [review]

**Pull request merged**